File: pe_4.hs
{-  Project Euler nubmer 4
A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.
Find the largest palindrome made from the product of two 3-digit numbers.
-}

{-  pseudo code
  
I want to do this with list comprehensions.

-}

module Main where
    m = show 41214    

    -- check if palindrome
    palindrome x = if ((show x) == reverse (show x)) then True else False
    
    allProducts = [i*j | i <- [100..999], j <- [100..999] ]

    --for me this is modified from "Learn you a Haskell"
    --though I've learned it's all over
    quicksort_descending :: (Ord a) => [a] -> [a]  
    quicksort_descending [] = []  
    quicksort_descending (x:xs) =   
        let smallerSorted = quicksort_descending [a | a <- xs, a <= x]  
            biggerSorted = quicksort_descending [a | a <- xs, a > x]  
        in  biggerSorted ++ [x] ++ smallerSorted
    
    -- helper function (wrapper?) to get solution
    -- takes the first item from sorted descending list of only the palidromes 
    -- of all products of three digit numbers 
    f = head (quicksort_descending (filter palindrome allProducts))

{-  RESULT

f produces 906606, correct answer

-}
    -- needs main value to compile
    main = undefined

File: pe_4.js
/*  Project Euler nubmer 4
A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

Find the largest palindrome made from the product of two 3-digit numbers.
*/

/*  pseudo code
  
  function isPalindrome returns boolean

  for i = 999 downto ...
     for j = i downto ...
         isPalindrome?
         break
*/

//Changed to true to skip the loops for debugging
let found = false;
let i = 999;
palindromes = [];
let highest = 1;

while (( !found) && (i > 99)) {
    let j = i;
    while ((!found) && (j > 99)) {
        test = i * j;
        if (isPalindrome(test)) {
            //found = true;
            if (test > highest) highest = test;
			if ((palindromes.length > 0) && (test > palindromes[0]))
			    palindromes.unshift(test);
			else
				palindromes.push(test);
        }
        j -= 1;
    }
    i -= 1;
}

console.log("Ahem.");
console.log(palindromes);
console.log("From highest ", highest, " and from the list ", palindromes[0]);
//let number = 9187;
//console.log("From main ", number);
/*while (number > 9000) {
	isPalindrome(number)
	number --;
}*/
function isPalindrome (x) {
    //convert number to string
    let string = x.toString();
    let backwards = string.split('').reverse().join('');
	 if (string == backwards) {
		  //console.log(string, " is palindrome!");
        return true;
	 }
	 else {
		  //console.log(string, " ain't!");
        return false
	 }
        
}